src/casp/services/api/data_manager/cell_operations.py

```diff
@@ -111,31 +112,72 @@ def get_neighborhood_distance_summary_dev_details(
 
-        query_job = self.block_coo_matrix_db_client.query(query=sql_query)
+        if len(cell_ids) > settings.MAX_CELL_IDS_PER_QUERY:
+            raise exceptions.CellMetadataDatabaseError(
+                f"Number of cell IDs exceeds the maximum of {settings.MAX_CELL_IDS_PER_QUERY}"
+            )
 
-        return bigquery_response_parsers.parse_get_cells_job(
-            query_job=query_job, cell_metadata_features=metadata_feature_names
-        )
+        with self.system_data_db_session.begin():
+            # Creating temp table, it will be dropped automatically when the session is closed
+            metadata = sa.MetaData()
+            cell_info_tmp_table = sa.Table(
+                "cells_cellinfotemptable", metadata, sa.Column("cas_cell_index", sa.Integer), prefixes=["temporary"]
+            )
+            metadata.create_all(self.system_data_db_session.bind)
+
+            # Insert values into the temp table
+            insert_query = cell_info_tmp_table.insert().values([{"cas_cell_index": cell_id} for cell_id in cell_ids])
+            self.system_data_db_session.execute(insert_query)
+
+            # Select values from the temp table, joining with the CellInfo table
+            select_columns = [getattr(models.CellInfo, feature_name) for feature_name in feature_names]
+            statement = sa.select(select_columns).join(
+                cell_info_tmp_table, models.CellInfo.cas_cell_index == cell_info_tmp_table.c.cas_cell_index
+            )
+
+            query_result = self.system_data_db_session.execute(statement).fetchall()
+
+        # Reconstruct the ordered list based on cell_ids
+        index_to_metadata = {row["cas_cell_index"]: row for row in query_result}
+        ordered_metadata = (index_to_metadata[cell_id] for cell_id in cell_ids if cell_id in index_to_metadata)
+
+        return [schemas.CellariumCellMetadata(**cell_metadata_row) for cell_metadata_row in ordered_metadata]
+
+    def get_cell_metadata_by_ids(
+        self, cell_ids: t.List[int], metadata_feature_names: t.List[str]
+    ) -> t.List[schemas.CellariumCellMetadata]:
+        """
+        Get cells by CAS IDs, maintaining the order of `cell_ids`. If the number of querying cell IDs exceeds the
+        maximum allowed, split the query into multiple queries. Database gets overwhelmed when querying a large number
+        of cell.
+
+        :param cell_ids: CAS cell IDs to retrieve metadata for.
+        :param metadata_feature_names: List of feature names to retrieve.
+
+        :return: List of dictionaries representing the query results, ordered according to `cell_ids`.
+        """
+        if len(cell_ids) >= settings.MAX_CELL_IDS_PER_QUERY:
+            neighbors_metadata = []
+            for i in range(0, len(cell_ids), settings.MAX_CELL_IDS_PER_QUERY):
+                neighbors_metadata += self._get_cell_metadata_by_ids(
+                    cell_ids=cell_ids[i : i + settings.MAX_CELL_IDS_PER_QUERY],
```

Comment:
Does it make sense to replace `i + settings.MAX_CELL_IDS_PER_QUERY` with `min(len(cell_ids), i + settings.MAX_CELL_IDS_PER_QUERY)`? If `cell_ids` is either a list or ndarray, we get the right behavior. If it is some other iterable that may behave differently if the upper limit of the slice is above the lenghth, we will be in trouble.